pkg/webhook/policy.go

```diff
@@ -624,13 +622,17 @@ func createReviewForResultant(obj *unstructured.Unstructured, ns *corev1.Namespa
 	}
 }
 
-func getViolationRef(gkNamespace, rkind, rname, rnamespace, ckind, cname, cnamespace string) *corev1.ObjectReference {
-	return &corev1.ObjectReference{
+func getViolationRef(enamespace, rkind, rname, rrv string, ruid types.UID) *corev1.ObjectReference {
+	ref := &corev1.ObjectReference{
 		Kind:      rkind,
 		Name:      rname,
-		UID:       types.UID(rkind + "/" + rnamespace + "/" + rname + "/" + ckind + "/" + cnamespace + "/" + cname),
```

Comment:
why do we want to remove constraint information here? I think that is useful to know.

Reply:
This information is still all contained in the events annotations so it is not lost :)
```golang
	annotations := map[string]string{
		"process":                    "audit",
		"auditTimestamp":             timestamp,
		logging.EventType:            "violation_audited",
		logging.ConstraintGroup:      constraint.GroupVersionKind().Group,
		logging.ConstraintAPIVersion: constraint.GroupVersionKind().Version,
		logging.ConstraintKind:       constraint.GetKind(),
		logging.ConstraintName:       constraint.GetName(),
		logging.ConstraintNamespace:  constraint.GetNamespace(),
		logging.ConstraintAction:     string(enforcementAction),
		logging.ResourceGroup:        resourceGroupVersionKind.Group,
		logging.ResourceAPIVersion:   resourceGroupVersionKind.Version,
		logging.ResourceKind:         resourceGroupVersionKind.Kind,
		logging.ResourceNamespace:    rnamespace,
		logging.ResourceName:         rname,
	}
```

The main reason to use the involved objects real UID would be so these events can be discovered by existing tools that link events to the resources they are about. For example the describer process in kubectl searches events in the namespace for matching `name` `kind` and `uid` and will display the events alongside the resource which is useful information for end users and brings greater visibility to possible issues. Below is what that might look like.

```sh
$ kubectl describe deployment grafana-operator-controller-manager
Name:               grafana-operator-controller-manager
Namespace:          offending-namespace
...
...
...
Events:
  Type     Reason          Age                From              Message
  ----     ------          ----               ----              -------
  Warning  AuditViolation  9m (x28 over 27h)  gatekeeper-audit  Constraint: must-have-probes, Message: [Implied by expand-deployments] Container <kube-rbac-proxy> in this <Pod> has no <readinessProbe> and is selected by service <grafana-operator-controller-manager-metrics-service> with targetPort(s) ["https"].
```

This of course only works if the events are generated in the involved objects namespace. If the arbitrary (current) UID remains events will not visible on objects when using kubectl or other tools that associate resources with events that are about them.

https://github.com/kubernetes/kubectl/blob/master/pkg/describe/describe.go#L280
https://github.com/kubernetes/kubectl/blob/master/pkg/describe/describe.go#L5444

Reply:
Thanks for sharing discoverability by existing tools. That makes sense. 

The UID is concatenated as such today is due to the concerns described [here](https://github.com/open-policy-agent/gatekeeper/pull/739#discussion_r459721241) for clobbering. Could the same concern be applicable here if there are multiple events generated for the same violating resource?

Similar to an earlier comment, for the case where the flag is not enabled, lets restore it back to what it was to reduce impact especially because we know clobbering could happen.